Add support for OpenELM

Blaizzy opened the pull request:
This PR adds support for openELM.
Twitter: @Prince_Canuma

Blaizzy commented:
@awni I couldn't replicate the qkv splitting logic from the transformers implementation.

```python
# [B, (q_h + k_h + v_h), S, h] --> [B, q_h, S h], [B, k_h, S, h], [B, v_h, S, h]
queries, keys, values = qkv.split(
      [self.num_q_heads, self.num_k_heads, self.num_v_heads], dim=1
)
```

Here is what I tried:
```
queries, kv = mx.split(qkv, [self.n_heads], axis=1)
keys , values = mx.split(kv, 2, axis=1)
```

The outputs don't yet make sense with this approach.

Goekdeniz-Guelmez commented:
have you tried splitting them in one go like: `queries, keys, values = mx.split(qkv, 3, axis=-1)`, why do you have to split them into query and kv and then the kv again into key and value pairs?

awni commented:
Closes #723

Blaizzy commented:
> @awni I couldn't replicate the qkv splitting logic from the transformers implementation.
> 
> 
> 
> ```python
> 
> # [B, (q_h + k_h + v_h), S, h] --> [B, q_h, S h], [B, k_h, S, h], [B, v_h, S, h]
> 
> queries, keys, values = qkv.split(
> 
>       [self.num_q_heads, self.num_k_heads, self.num_v_heads], dim=1
> 
> )
> 
> ```
> 
> 
> 
> Here is what I tried:
> 
> ```python
> 
> queries, kv = mx.split(qkv, [self.n_heads], axis=1)
> 
> keys , values = mx.split(kv, 2, axis=1)
> 
> ```
> 
> and
> 
> ```python
> 
> queries = qkv[:, :self.n_heads, :, :]
> 
> kv = qkv[:, self.n_kv_heads:, :, :]
> 
> keys, values = kv
> 
> ```
> 
> 
> 
> However, the outputs don't yet make sense with this approach.


@awni the issue still remains. 

The model loads well but produces nonsensical responses.

I think mx.split works differently from the one they used. 

I'm trying to think of different options but I'm bit stuck.

awni commented:
You should use split in one go rather than calling it multiple times. See the [docs here](https://ml-explore.github.io/mlx/build/html/python/_autosummary/mlx.core.split.html#mlx.core.split)

As for why the model doesn't generate good output, I'm not sure there. I haven't looked at it in detail. Maybe double check the reference implementation? A good next step if you are stuck is to

1. step through the reference implementation and make sure both models are generating the same outputs up to the same point.
2. Once you find the point at which the outputs differ, that should help you diagnose the problem in the code.

Blaizzy commented:
@awni thank you very for the advice!

Thank God MLX has similar tensor/array slicing as torch and numpy!

I knew the error was in the kv slicing. I just needed to sitdown and focus :)

It's done now ✅

awni commented on llms/mlx_lm/tuner/utils.py:
```diff
@@ -88,6 +88,8 @@ def linear_to_lora_layers(
         keys = set(["att_proj"])
     elif model.model_type == "phi3":
         keys = set(["self_attn.qkv_proj"])
+    elif model.model_type == "openelm":
```
```suggestion
    elif model.model_type in ["phi3", "openelm"]:
        keys = set(["self_attn.qkv_proj"])
```

Blaizzy replied:
Done ✅

awni commented:
Looks great, thanks for the addition! Could you address the two comments I left, then we can merge?

Blaizzy commented on llms/mlx_lm/models/openelm.py:
```diff
@@ -0,0 +1,249 @@
+        ).transpose(0, 2, 1, 3)
+
+        # [B, (q_h + k_h + v_h), S, h] --> [B, q_h, S h], [B, k_h, S, h], [B, v_h, S, h]
+        queries = qkv[:, : self.n_heads, :, :]
+        keys = qkv[:, self.n_heads : (self.n_heads + self.n_kv_heads), :, :]
+        values = qkv[:, (self.n_heads + self.n_kv_heads) :, :, :]
```
I was looking for exactly that. The MLX.core documentation doesn't explain how it works well, perhaps  some examples would help.

Blaizzy replied:
It's done ✅ 

Also, I ran some tests and there is little difference in speed.

Blaizzy commented:
> Looks great, thanks for the addition! Could you address the two comments I left, then we can merge?

Most welcome! This was an interesting architecture.

The comments are resolved ✅

awni commented on llms/mlx_lm/models/openelm.py:
```diff
@@ -0,0 +1,247 @@
+            if not all(key in self.rope_scaling for key in required_keys):
+                raise ValueError(f"rope_scaling must contain keys {required_keys}")
+
+            if self.rope_scaling["type"] != "linear":
+                print(
+                    "[WARNING] rope_scaling 'type' currently only supports 'linear' setting rope scaling to false."
+                )
```
What is the purpose of this? I don't think this model uses rope scaling, so could you remove the stuff on rope scaling and just default it to 1.0?

Blaizzy replied:
Yes, will do !

I thought the logic defaults to 1 anyways.

Blaizzy replied:
Done ✅

awni commented:
Do you mind to run the pre-commit hooks? Looks like the linter failed.

Blaizzy commented:
Yeah, it's because I removed scale argment.

It's done ✅

awni reviewed (approved):
Thanks!!
